### PR TITLE
Make repair/rearm of ground units queueable

### LINKED ISSUE #16462: Repairable activity management is bogus
As noted in #16352, `Repairable` badly manages the activity queue when sending units to repair, which causes bogus behaviour when other activities are queued after the repair. Most of this logic should move inside `Resupply`, using child activities.

> The issue that if you sent a unit to repair itself and immediately queue another command afterwards, that second command will be executed after the unit reaches the repair pad. After completing that order it will then drive back to the repair pad and execute the AfterReachActivities. The cause is that the latter get queued from a CallFunc. This was done to enable carryalls carrying units to repair pads.

### LINKED ISSUE #16463: Aircraft remain on the service depot after repair
Repro case on bleed: damage a mig and order it to repair at the service depot. Wait for it to finish repairing.

Expected Behaviour: It should take off immediate after repairing (like vehicles) or at worst remain on the ground until another unit needs to repair (like the airfield).

Actual result: It remains on the depot, blocking other units from repairing, until it is manually ordered to leave.

#16352 does not appear to have affected this.

### LINKED ISSUE #16492: Unexpected behavior when interrupting rearming
See gif below.

<details>

landingbroken

Note that rearming is aborted when giving the move order.

</details>

This does not happen with helicopters.

### LINKED ISSUE #16761: Service Depot bug: new rallypoint movecmd even thought other movecmd exists
After a vehicle is repaired on the service depot a new move command is added to the bottom of the units move cmd list, even if it already has other commands.
This results in vehicles moving back to the rallypoint (after individual move cmds where executed).

Tested with TS service depot and RA service depot on bleed.

Regression [PR #13093](https://github.com/OpenRA/OpenRA/pull/13093)
openra_bug_servicedepot

### PR DESCRIPTION
**Pull request by reaperrr**
Moves `Repairable(Near)` activity handling to `Resupply` activity, plus a few fixes.

Implements first step of https://github.com/OpenRA/OpenRA/issues/15023#issuecomment-489332380

Fixes #16462.
Fixes #16463.
Fixes #16761.
Fixes #16492.

### CODE REVIEW

File: OpenRA.Mods.Common/Traits/Repairable.cs [hunk 1]
```diff
@@ -123,41 +123,19 @@ void IResolveOrder.ResolveOrder(Actor self, Order order)
...
 					self.CancelActivity();
 
-				self.SetTargetLine(order.Target, Color.Green);
-				var activities = ActivityUtils.SequenceActivities(self,
-					movement.MoveToTarget(self, order.Target, targetLineColor: Color.Green),
-					new CallFunc(() => AfterReachActivities(self, order, movement)));
-
-				self.QueueActivity(new WaitForTransport(self, activities));
-				TryCallTransport(self, order.Target, new CallFunc(() => AfterReachActivities(self, order, movement)));
+				self.QueueActivity(new MoveToResupplier(self, order.Target));
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Repairable.cs]: We really want to be queuing `Resupply` here, and that checks whether it is in the correct position and moves if needed. See https://github.com/OpenRA/OpenRA/blob/bleed/OpenRA.Mods.Common/Activities/HarvestResource.cs#L62 for a recent example

**Comment by reaperrr** [comment]
Updated.

**Comment by teinarss** [comment]
We are missing the target lines when ordering repairing. Is this intended and will be fixed with #16549?

**Comment by reaperrr** [comment]
That wasn't intended, no.

Updated.

**Review by teinarss — DISMISSED** [review]
LGTM

**Review by abcdefg30 — DISMISSED** [review]
Helicopters in RA don't take off at all if you queue a reload and repair activity.

**Comment by reaperrr** [comment]
> Helicopters in RA don't take off at all if you queue a reload and repair activity.

Not this PR's fault and intentional, see discussion in #16660. We can still change that back after the first playtest if the feedback on that is negative.

As far as this PR is concerned, manually set `AbortOnResupply: false` if you need a test case for aircraft (works fine according to my testing).

**Comment by reaperrr** [comment]
As discussed on IRC, fixing that in a non-work-around way requires #16509 to be finished and merged first.

**Comment by pchote** [comment]
#16509 has been merged.

**Comment by pchote** [comment]
> 23:24 <+pchote> re the comments in #16721, I have a sinking feeling that we may end up having to rework AbortOnResupply before the end of the playtest series
23:24 <@orabot> Pull request #16721 (open) by reaperrr: Make repair/rearm of ground units queueable | http://bugs.openra.net/16721
23:25 <+reaperrr> ugh but yeah, am not that surprised
23:25 <+pchote> if you have a damaged aircraft and want to queue a repair/reload and then move somewhere that isn't the rallypoint then this really should be possible
23:25 <+pchote> and this is independent from whether they break off after a single attack run
23:26 <+pchote> tovl reworks the flag a bit in #16481
23:26 <@orabot> Pull request #16481 (open) by tovl: Move ChildActivity handling into base Activity class | http://bugs.openra.net/16481
23:27 <+pchote> I expect that after that has been merged that we'll need to take it further and probably move it to AttackAircraft
23:27 <+pchote> have it abort FlyAttack only, not the whole quue
23:27 <+pchote> *queue
23:28 <+reaperrr> yeah

**Comment by reaperrr** [comment]
Updated. Fingers crossed that the last commit might already be enough.

**Comment by pchote** [comment]
Needs a rebase after #16481 was merged.

**Comment by reaperrr** [comment]
Rebased.

File: OpenRA.Mods.Common/Activities/Air/FlyAttack.cs [hunk 2]
```diff
@@ -111,11 +111,17 @@ public override bool Tick(Actor self)
 				return true;
 			}
 
-			// If all valid weapons have depleted their ammo and Rearmable trait exists, return to RearmActor to reload and then resume the activity
+			// If all valid weapons have depleted their ammo and Rearmable trait exists, return to RearmActor to reload
 			if (rearmable != null && !useLastVisibleTarget && attackAircraft.Armaments.All(x => x.IsTraitPaused || !x.Weapon.IsValidAgainst(target, self.World, self)))
 			{
-				QueueChild(new ReturnToBase(self));
-				return aircraft.Info.AbortOnResupply;
+				// End this activity after reload has been queued if AbortOnResupply is true,
```
Comment (pchote) [inline OpenRA.Mods.Common/Activities/Air/FlyAttack.cs]: This is what the `return attackAircraft.Info.AbortOnResupply;` should already be doing. Are you sure these changes are still needed after the rebase?

Comment (reaperrr) [reply OpenRA.Mods.Common/Activities/Air/FlyAttack.cs]: No, but I also wasn't sure what happens to the child when the parent activity ends. It seems to work as intended though, so I'll drop my change.

**Comment by reaperrr** [comment]
Updated. Also added an update rule for mods that intentionally set `AbortOnResupply` to `false`.

**Comment by pchote** [comment]
Needs a rebase again / still?

File: OpenRA.Mods.Common/Activities/Resupply.cs [hunk 3]
```diff
@@ -56,46 +61,77 @@ public Resupply(Actor self, Actor host, WDist closeEnough)
...
+				{
+					aircraft.AllowYieldingReservation();
+					if (aircraft.Info.TakeOffOnResupply)
+						Queue(new TakeOff(self));
 
-			if (IsCanceling || host.Type == TargetType.Invalid
-				|| (closeEnough.LengthSquared > 0 && !host.IsInRange(self.CenterPosition, closeEnough)))
+					return true;
+				}
+				else if (self.Info.HasTraitInfo<MobileInfo>())
```
Comment (pchote) [inline OpenRA.Mods.Common/Activities/Resupply.cs]: Can save some trait(info) lookups here by doing
```
var move = self.Trait<IMove>();
var aircraft = move as Aircraft;
```

above, and then using `move` instead of `MobileInfo` and the lookups below.

File: OpenRA.Mods.Common/Activities/Resupply.cs [hunk 4]
```diff
@@ -56,46 +61,77 @@ public Resupply(Actor self, Actor host, WDist closeEnough)
...
+			}
+			else if (IsCanceling || host.Type != TargetType.Actor || !host.Actor.IsInWorld || host.Actor.IsDead)
 			{
 				// This is necessary to ensure host resupply actions (like animations) finish properly
 				foreach (var notifyResupply in notifyResupplies)
 					notifyResupply.ResupplyTick(host.Actor, self, ResupplyType.None);
 
 				return true;
 			}
+			else if (activeResupplyTypes != 0 && !self.Info.HasTraitInfo<AircraftInfo>() &&
```
Comment (pchote) [inline OpenRA.Mods.Common/Activities/Resupply.cs]: can reuse `aircraft` instead of querying `AircraftInfo`

File: OpenRA.Mods.Common/Activities/Resupply.cs [hunk 5]
```diff
@@ -112,6 +148,16 @@ public override bool Tick(Actor self)
 				if (aircraft != null)
 					aircraft.AllowYieldingReservation();
 
+				var actorHasNextActivity = self.CurrentActivity.NextActivity != null;
+				if (!actorHasNextActivity && repairableNear == null)
+				{
+					var rp = host.Actor.TraitOrDefault<RallyPoint>();
+					if (rp != null)
+						Queue(self.Trait<IMove>().MoveTo(rp.Location, host.Actor));
```
Comment (pchote) [inline OpenRA.Mods.Common/Activities/Resupply.cs]: and then reuse `move` here.

File: OpenRA.Mods.Common/Traits/Repairable.cs [hunk 6]
```diff
@@ -99,65 +97,44 @@ bool CanRearm()
...
-					return;
+			// Repair orders are only valid for own/allied actors,
+			// which are guaranteed to never be frozen.
+			if (order.Target.Type != TargetType.Actor)
+				return;
 
-				if (!CanRepairAt(order.Target.Actor) || (!CanRepair() && !CanRearm()))
-					return;
+			// Aircraft handle Repair orders directly in the Aircraft trait
+			if (self.Info.HasTraitInfo<AircraftInfo>())
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Repairable.cs]: You could cache a `bool isAircraft` from `INotifyCreated.Created` above.

...or maybe remove this Aircraft special case completely, like we discussed the other day? I assume that didn't go so well...

Comment (reaperrr) [reply OpenRA.Mods.Common/Traits/Repairable.cs]: > I assume that didn't go so well...

Nope, another can of worms. Going by experience, this looks like something that could easily take 4-8 weeks to get right (delays in reviewing and updating calculated in).

**Comment by reaperrr** [comment]
> Needs a rebase again / still?

Again. The FlightDynamics PR was merged after I had rebased this one.

Updated and rebased.

**Comment by pchote** [comment]
The " Fix Resupply regression" commit references `AircraftInfo.TakeOffOnResupply` instead of `AircraftInfo.FlightDynamics`. Can you please fix that commit so it compiles cleanly for the sake of future bisects?

File: OpenRA.Mods.Common/Activities/Resupply.cs [hunk 7]
```diff
@@ -119,10 +153,19 @@ public override bool Tick(Actor self)
 
 			if (activeResupplyTypes == 0)
 			{
-				var aircraft = self.TraitOrDefault<Aircraft>();
 				if (aircraft != null)
 					aircraft.AllowYieldingReservation();
 
+				var actorHasNextActivity = self.CurrentActivity.NextActivity != null;
+				if (!actorHasNextActivity && repairableNear == null)
```
Comment (pchote) [inline OpenRA.Mods.Common/Activities/Resupply.cs]: TIL that `RepairableNear` was never hooked up to rally points. Can we please fix this while we are here?

**Review by pchote — COMMENTED** [review]
One more request, then LGTM

File: OpenRA.Mods.Common/Activities/Resupply.cs (hunk 7 above)
Comment (reaperrr) [reply OpenRA.Mods.Common/Activities/Resupply.cs]: Personally, I found the move to rallypoint a bit annoying for RepairableNear, because
- the repaired actor isn't blocking the only repair slot if it doesn't move
- depending on where it was repaired, it may have to move around the entire building to reach the rallypoint
- as ship yards/sub pens are also production facilities, things can get crowded at the rp even faster if repaired actors move there as well

I don't really mind changing it anyway, just wanted to point this out.

**Comment by reaperrr** [comment]
Updated.

> Can you please fix that commit so it compiles cleanly for the sake of future bisects?

Done.

**Comment by fruestueck** [comment]
Fixes #16761 

I had a small bug where a vehicle first moved to the top-left cell obove the RA service depot but I can't reproduce it. Could be nothing.

**Review by abcdefg30 — DISMISSED** [review]
Something seems to be going on with `RepairableNear`:
danceDestroy
Testcase: Issue a stop order while repairing. The unit will not respond to any other orders.

Otherwise the PR looks good to me.

**Comment by reaperrr** [comment]
Fixed.
The problem was that the `MoveToTarget` on cancellation is bogus for `RepairableNear` because a) it doesn't need to leave the host and b) it would always remain within `closeEnough` range, re-queueing the move infinitely.

Also merged the fix and the previously separate `RepairableNear` commit into 
4984788.

**Comment by pchote** [comment]
Unfortunately `RepairableNear` is still not behaving correctly.

Testcase:
1) Damage a boat
2) order it to repair + queue a move somewhere afterwards.
3) issue a non-queued move order while it is repairing.

Expected behaviour: repair and queued move are cancelled, boat moves immediately to ordered location.
Actual behaviour: repair continues until completed, boat then moves to the original queued location.

**Comment by reaperrr** [comment]
I have no friggin' clue why. The way I understand the updated Activity Tick code, `return true` should be equal to the old `return NextActivity`, and `Mobile` calls `self.CancelActivity()` on non-queued move orders, so it doesn't make any sense to me that the resupply continues, nor that the queued orders aren't cancelled.

**Comment by pchote** [comment]
An important detail from `Activity.Tick` is
https://github.com/OpenRA/OpenRA/blob/1c03fb9e516e0690a2008749d754ad453e0dd733/OpenRA.Game/Activities/Activity.cs#L106-L108

The `Resupply` activity has a custom `Cancel` override that prevents `base.Cancel` (and therefore the child activities) from being cancelled if there are queued activities. This forces the child activities to be uncancellable and run until completion.

**Comment by tovl** [comment]
The `Cancel` override needs to be removed. I don't even remember what I was thinking there. I guess this was supposed to help with the issue of `Resupply` being cancelled by a stop command, but it is complete bogus.

**Comment by reaperrr** [comment]
Updated, left the updates in separate commits (for now).

File: OpenRA.Mods.Common/Activities/Resupply.cs [hunk 8]
```diff
@@ -56,46 +65,86 @@ public Resupply(Actor self, Actor host, WDist closeEnough)
...
 				foreach (var notifyResupply in notifyResupplies)
 					notifyResupply.ResupplyTick(host.Actor, self, ResupplyType.None);
 
-				return false;
-			}
+				if (aircraft != null)
+				{
+					aircraft.AllowYieldingReservation();
+					if (aircraft.Info.FlightDynamics.HasFlag(FlightDynamic.TakeOffOnResupply))
+						Queue(new TakeOff(self));
```
Comment (tovl) [inline OpenRA.Mods.Common/Activities/Resupply.cs]: `ReturnToBase` already takes care of queueing the `TakeOff` so this seems a bit redundant. Also, is there a reason why this is not queued as a childactivity? This wouldn't work correctly when other activities are queued after `Resupply.`

Comment (pchote) [reply OpenRA.Mods.Common/Activities/Resupply.cs]: I've lost track of the specifics about `ReturnToBase` and taking off, but having the `TakeOff` `Queue`d after other user-ordered activities is definitely an issue.

**Comment by reaperrr** [comment]
Updated.

File: OpenRA.Mods.Common/Activities/Resupply.cs [hunk 9]
```diff
@@ -78,18 +78,29 @@ public override bool Tick(Actor self)
 				if (aircraft != null)
 				{
 					aircraft.AllowYieldingReservation();
-					if (aircraft.Info.FlightDynamics.HasFlag(FlightDynamic.TakeOffOnResupply))
-						Queue(new TakeOff(self));
 
-					return true;
+					// ReturnToBase only handles TakeOffOnResupply when Resupply finishes, not when it's cancelled prematurely
+					if (aircraft.Info.FlightDynamics.HasFlag(FlightDynamic.TakeOffOnResupply))
```
Comment (tovl) [inline OpenRA.Mods.Common/Activities/Resupply.cs]: The `TakeOff` should be suppressed when `Resupply` is called by `ReturnToBase` with `alwaysLand = true`. This can be done by adding an `alwaysStay` parameter to the constructor and having `ReturnToBase` pass `alwaysLand` to that.

File: OpenRA.Mods.Common/Activities/Resupply.cs [hunk 10]
```diff
@@ -148,16 +159,16 @@ public override bool Tick(Actor self)
 
 			if (activeResupplyTypes == 0)
 			{
+				// Resupply properly finished, so ReturnToBase will handle the TakeOffOnResupply
```
Comment (tovl) [inline OpenRA.Mods.Common/Activities/Resupply.cs]: `TakeOffOnResupply` is now handled in two different places. It might be better to remove the `TakeOff` handling from `ReturnToBase` and handle it here. I moved the `TakeOff` to `ReturnToBase` to handle the `alwaysLand` case, but that can be handled by passing the boolean down to `Resupply` instead (per the above comment).

File: OpenRA.Mods.Common/Activities/Resupply.cs [hunk 11]
```diff
@@ -78,18 +78,29 @@ public override bool Tick(Actor self)
 				if (aircraft != null)
```
Comment (tovl) [inline OpenRA.Mods.Common/Activities/Resupply.cs]: The entire block handling reservation, takeoff and rallypoints is now repeated twice, so it would make sense to pull this into a method so as to not repeat ourselves.

File: OpenRA.Mods.Common/Activities/Resupply.cs [hunk 12]
```diff
@@ -78,18 +78,29 @@ public override bool Tick(Actor self)
...
-				else if (repairableNear != null)
-					return true;
 				else
 				{
-					QueueChild(move.MoveToTarget(self, host));
-					return false;
+					// If there's no next activity, move to rallypoint if available, else just leave host if Repairable.
+					// Do nothing if RepairableNear (RepairableNear actors don't enter their host and will likely remain within closeEnough).
+					// If there's a next activity and we're not RepairableNear, first leave host if the next activity is not a Move.
+					if (self.CurrentActivity.NextActivity == null)
```
Comment (tovl) [inline OpenRA.Mods.Common/Activities/Resupply.cs]: It would be nice if ground units also had the option to stay on the resupplier after resupplying. This could be done by replacing `TakeOffOnResupply` with `LeaveOnRepair` and `LeaveOnRearm` which would be parameters of the `Repairable` and `Rearmable` trait respectively. This would be a step in separating the resupply logic from `Aircraft`. Having separate booleans for repair and rearm has the added benefit of allowing aircraft to takeoff after repairing on the repair pad but not after rearming on the airfield.

Comment (reaperrr) [reply OpenRA.Mods.Common/Activities/Resupply.cs]: I agree, but I'd like to leave that step for a follow-up, we're already quite late with the playtest.

Comment (tovl) [reply OpenRA.Mods.Common/Activities/Resupply.cs]: Yes, this is more something for a follow-up than something that needs to happen right now.

**Comment by reaperrr** [comment]
Updated.

File: OpenRA.Mods.Common/Activities/Resupply.cs [hunk 13]
```diff
@@ -148,22 +137,39 @@ public override bool Tick(Actor self)
...
+			return false;
+		}
+
+		void OnResupplyEnding(Actor self)
+		{
+			if (aircraft != null)
+			{
+				aircraft.AllowYieldingReservation();
+
+				// ReturnToBase only handles TakeOffOnResupply when Resupply finishes, not when it's cancelled prematurely
```
Comment (tovl) [inline OpenRA.Mods.Common/Activities/Resupply.cs]: This comment can now be removed.

Comment (reaperrr) [reply OpenRA.Mods.Common/Activities/Resupply.cs]: Done.

**Comment by tovl** [comment]
This does not seem to fix #16463.

**Comment by reaperrr** [comment]
> This does not seem to fix #16463.

Right, forgot that this would require more changes.
I've added a hacky short-term fix for the upcoming release, let's leave the proper fix (`TakeOffOn` flag list?) for a follow-up.

**Comment by tovl** [comment]
The fix works. However, aircraft with `TakeOffOnResupply == false` will not move to the rally point when taking off from the repair bay.

I also noticed that when an aircraft with `TakeOffOnResupply == true` is made to resupply using the `force-enter` command and the `stop` command is given it will forget to supress the `TakeOff`. This is due to the `Resupply` activity being reqeued without the `stayOnResupplier` parameter.

**Comment by reaperrr** [comment]
> The fix works. However, aircraft with TakeOffOnResupply == false will not move to the rally point when taking off from the repair bay.

`Aircraft.GetActorBelow()` only looks for Reservable actors, and Reservable in its current form won't mesh with repair-only hosts. I'd like to leave this as "known issue, but not serious enough to be a release blocker" and adress it properly during the next release cycle.

> I also noticed that when an aircraft with TakeOffOnResupply == true is made to resupply using the force-enter command and the stop command is given it will forget to supress the TakeOff. This is due to the Resupply activity being reqeued without the stayOnResupplier parameter.

Fixing this properly would likely require to remove the hacky re-queueing of `Resupply`, which in turn would require `Attack*` traits to only cancel their own activities but not the entire queue on "Stop" orders, which would require us to refactor `Actor.CancelActivity()`.
tl;dr: Another issue I'd like to defer to Next+1.

**Comment by tovl** [comment]
I agree with deferring these issues to Next+1.

As far as refactoring `Actor.CancelActivity()` is concerned, I already have a potential solution in mind that would not require us touching stop orders or requeueing `Resupply` but would require reworking `Actor.QueueActivity()`. I will try that out as soon as this is merged.

**Comment by tovl** [comment]
Found another bug: In D2K, orders queued after the repair order get ignored when the unit has been carried by a carryall.

**Comment by reaperrr** [comment]
I'm pretty sure it was that way before this PR already.
I could pass a sequence of `Resupply, NextActivity` instead of only `Resupply` through `RequestTransport`, not sure if that also implicitly passes any activities queued after `NextActivity`, though.

**Comment by tovl** [comment]
Any activity in the queue carries a reference to the next activity in the queue. So passing `NextActivity` would also implicitly pass `NextActivity.NextActivity`, `NextActivity.NextActivity.Nextactivity`, etc..., so I suspect this will work.

**Comment by reaperrr** [comment]
OK, not that easy, unfortunately. Resupply triggers the request right on its first tick, so any activity queued after the transport was requested is ignored. I think this might require refactoring how RequestTransport works, but that would be out of scope for this PR.

**Comment by pchote** [comment]
>  I think this might require refactoring how RequestTransport works, but that would be out of scope for this PR.

When the time comes, this can be done the same way I was arguing the transport locking for passengers should be done. Right now the carryall makes the requesting actor cancel its activity and queues a `WaitFor` until it has picked it up. Instead, `Carryable` should grant a condition that pauses `Mobile` so that the activity queue doesn't need to be touched at all.

**Comment by reaperrr** [comment]
Can we at least leave that to a follow-up? I'm literally getting a headache right now so I won't get that done today, and as long as this PR drags on I won't make progress on #16695's update rule, either.

**Comment by pchote** [comment]
"When the time comes" are the magic words :smile: IMO that is a job for Next + 1.

**Comment by reaperrr** [comment]
I was referring to the "right now" as well ;)

**Review by tovl — DISMISSED** [review]
With those known issues deferred to the future, I think this is good to go. :+1:

File: OpenRA.Mods.Common/Activities/Resupply.cs [hunk 14]
```diff
@@ -33,6 +33,7 @@ public class Resupply : Activity
 		readonly IMove move;
 		readonly Aircraft aircraft;
 		readonly bool stayOnResupplier;
+		readonly bool wasRepair;
```
Comment (teinarss) [inline OpenRA.Mods.Common/Activities/Resupply.cs]: nitpick: wasRepaired

**Review by teinarss — COMMENTED** [review]
Otherwise it looks good. Did not test the update rules

**Comment by reaperrr** [comment]
Fixed.

**Review by pchote — APPROVED** [review]
The important issues from earlier reviews appear to have been resolved.

The logic (both code and expected gameplay behaviours) here is complicated, so I expect that more regressions will probably be found in this code. The changes here are a needed step forward, so lets get these in as a first step to all the other fixes discussed above, and any other fixes that might be needed for Next.

**Comment by pchote** [comment]
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/bf379b06d37d013355fb28563aaa4ce4263b801b
